Api/Services/PaymentService.cs

```diff
@@ -0,0 +1,104 @@
+/// <param name="bankService"></param>
+/// <param name="walletService"></param>
+public class PaymentService(
+    ApiDbContext context,
+    BankService bankService,
+    WalletService walletService)
+{
+    /// <summary>
+    /// Function for paying the reservation deposit
+    /// </summary>
+    /// <param name="user"></param>
+    /// <param name="visitId"></param>
+    /// <returns></returns>
+    [ErrorCode(null, ErrorCodes.NotFound)]
+    [ErrorCode(null, ErrorCodes.AccessDenied)]
+    [ErrorCode(null, ErrorCodes.DepositFreeVisit)]
+    [ErrorCode(null, ErrorCodes.DepositAlreadyMade)]
+    [ErrorCode(null, ErrorCodes.InsufficientFunds)]
+    public async Task<Result<TransactionVM>> PayDepositAsync(User user, int visitId)
+    {
+        var visit = await context.Visits
+            .Where(v => v.VisitId == visitId)
+            .Include(v => v.Reservation)
+            .Include(v => v.Restaurant)
+            .FirstOrDefaultAsync();
+        //check if the visit exists
+        if (visit == null)
+        {
+            return new ValidationFailure
+            {
+                ErrorCode = ErrorCodes.NotFound,
+                ErrorMessage = ErrorCodes.NotFound,
+                PropertyName = null
+            };
+        }
+        //check if the visit belongs to the user
+        if (visit.ClientId != user.Id)
+        {
+            return new ValidationFailure
+            {
+                ErrorCode = ErrorCodes.AccessDenied,
+                ErrorMessage = ErrorCodes.AccessDenied,
+                PropertyName = null
+            };
+        }
+        //check if reservation requires a deposit
+        if (visit.Reservation!.Deposit is null)
+        {
+            return new ValidationFailure
+            {
+                ErrorCode = ErrorCodes.DepositFreeVisit,
+                ErrorMessage = ErrorCodes.DepositFreeVisit,
+                PropertyName = null
+            };
+        }
+        //check if the deposit was already paid
+        if (visit.Reservation!.DepositPaymentTime is not null)
+        {
+            return new ValidationFailure
+            {
+                ErrorCode = ErrorCodes.DepositAlreadyMade,
+                ErrorMessage = ErrorCodes.DepositAlreadyMade,
+                PropertyName = null
+            };
+        }
+        //check for sufficient funds to cover the transaction
+        var balance = await context.PaymentTransactions
+            .Where(p => p.UserId == user.Id)
+            .SumAsync(p => p.Amount);
+        if (balance < visit.Reservation.Deposit)
+        {
+            return new ValidationFailure
+            {
+                ErrorCode = ErrorCodes.InsufficientFunds,
+                ErrorMessage = ErrorCodes.InsufficientFunds,
+                PropertyName = null
+            };
+        }
+        bankService.SendMoneyToRestaurantAsync(visit.Restaurant, visit.Reservation.Deposit);
+        var transaction = await walletService.PayDepositAsync(user, visit);
```

Comment:
Tutaj uwaga, w przykładowym kawałku napisałem `WalletService.Debit`. Wiedza o tym, że to jest opłata kaucji, nie jest odpowiedzialnością WalletService. On po prostu odpowiada za obniżenie stanu konta. (upuściłem że ta metoda powinna przyjmować tytuł transakcji)